cpp/src/groupby/sort/group_dense_rank_scan.cu

```diff
@@ -0,0 +1,81 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+std::unique_ptr<column> dense_rank_scan(table_view const& order_by,
+                                        cudf::device_span<size_type const> group_labels,
+                                        cudf::device_span<size_type const> group_offsets,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::mr::device_memory_resource* mr)
+{
+  auto d_order_by    = table_device_view::create(order_by);
+  auto ranks         = make_fixed_width_column(cudf::data_type{cudf::type_to_id<size_type>()},
+                                       order_by.num_rows(),
+                                       mask_state::ALL_VALID,
+                                       stream,
+                                       mr);
+  auto mutable_ranks = ranks->mutable_view();
+
+  if (has_nulls(order_by)) {
+    row_equality_comparator<true> row_comparator(*d_order_by, *d_order_by, true);
+    thrust::tabulate(rmm::exec_policy(stream),
+                     mutable_ranks.begin<size_type>(),
+                     mutable_ranks.end<size_type>(),
+                     [row_comparator,
+                      labels  = group_labels.data(),
+                      offsets = group_offsets.data()] __device__(size_type row_index) {
+                       return (row_index == offsets[labels[row_index]] ||
+                               !row_comparator(row_index, row_index - 1))
+                                ? 1
+                                : 0;
+                     });
+  } else {
+    row_equality_comparator<false> row_comparator(*d_order_by, *d_order_by, true);
+    thrust::tabulate(rmm::exec_policy(stream),
+                     mutable_ranks.begin<size_type>(),
+                     mutable_ranks.end<size_type>(),
+                     [row_comparator,
+                      labels  = group_labels.data(),
+                      offsets = group_offsets.data()] __device__(size_type row_index) {
+                       return (row_index == offsets[labels[row_index]] ||
+                               !row_comparator(row_index, row_index - 1))
+                                ? 1
+                                : 0;
+                     });
+  }
+
+  thrust::inclusive_scan_by_key(rmm::exec_policy(stream),
+                                group_labels.begin(),
+                                group_labels.end(),
+                                mutable_ranks.begin<size_type>(),
+                                mutable_ranks.data<size_type>());
```

Comment:
`begin()` and `data()` seem to be the same. Is this wrong? Should it be something else?

Reply:
They will be the same, switched to using `begin()` for both. `inclusive_scan_by_key` supports writing back to the data location and it voids having to allocate an intermediate column.